src/Development/IDE/Core/Shake.hs

```diff
@@ -379,57 +403,126 @@ shakeProfile :: IdeState -> FilePath -> IO ()
+--   Progress is reported only on the system actions.
+shakeRestart :: IdeState -> [Action ()] -> IO ()
+shakeRestart it@IdeState{shakeExtras=ShakeExtras{logger}, ..} systemActs =
     withMVar'
-        shakeAbort
-        (\stop -> do
-              (stopTime,_) <- duration stop
-              logDebug logger $ T.pack $ "Starting shakeRun (aborting the previous one took " ++ showDuration stopTime ++ ")"
+        shakeSession
+        (\runner -> do
+              (stopTime,queue) <- duration (cancelShakeSession runner)
+              logDebug logger $ T.pack $ "Restarting build session (aborting the previous one took " ++ showDuration stopTime ++ ")"
+              return queue
         )
         -- It is crucial to be masked here, otherwise we can get killed
-        -- between spawning the new thread and updating shakeAbort.
+        -- between spawning the new thread and updating shakeSession.
         -- See https://github.com/digital-asset/ghcide/issues/79
-        (do
-              start <- offsetTime
-              aThread <- asyncWithUnmask $ \restore -> do
-                   res <- try (restore $ shakeRunDatabaseProfile shakeProfileDir shakeDb acts)
-                   runTime <- start
-                   let res' = case res of
-                            Left e -> "exception: " <> displayException e
-                            Right _ -> "completed"
-                       profile = case res of
-                            Right (_, Just fp) ->
-                                let link = case filePathToUri' $ toNormalizedFilePath' fp of
-                                                NormalizedUri _ x -> x
-                                in ", profile saved at " <> T.unpack link
-                            _ -> ""
-                   let logMsg = logDebug logger $ T.pack $
-                        "Finishing shakeRun (took " ++ showDuration runTime ++ ", " ++ res' ++ profile ++ ")"
-                   return (fst <$> res, logMsg)
-              let wrapUp (res, _) = do
-                    either (throwIO @SomeException) return res
-              _ <- async $ do
-                  (_, logMsg) <- wait aThread
-                  logMsg
-              pure (cancel aThread, wrapUp =<< wait aThread))
+        (fmap (,()) . newSession it systemActs)
+
+-- | Enqueue an action in the existing 'ShakeSession'.
+--   Returns a computation to block until the action is run, propagating exceptions.
+--   Assumes a 'ShakeSession' is available.
+--
+--   Appropriate for user actions other than edits.
+shakeEnqueue :: IdeState -> Action a -> IO (IO a)
+shakeEnqueue IdeState{shakeSession} act =
+    withMVar shakeSession $ \s -> runInShakeSession s act
+
+-- Set up a new 'ShakeSession' with a set of initial system and user actions
+-- Will crash if there is an existing 'ShakeSession' running.
+-- Progress is reported only on the system actions.
+-- Only user actions will get re-enqueued
```

Comment:
Not sure this is the right way around. It is really the system actions we want to requeue in the end I think. This distinction seems premature anyway.

Reply:
'system' and 'user' actions are not the best names, but are the best I could come up late at night, sorry about that.

- system actions: typechecks. We don't want to reenqueue typechecks since a fresh set of them are always provided when restarting the session.

- user actions: everything else. They must always be completed (unless the lsp-client cancels them) or the `clientMsgChan` thread will deadlock.